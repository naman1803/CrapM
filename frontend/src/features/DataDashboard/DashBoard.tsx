import { Box, Divider, Grid2, Typography } from '@mui/material';
import React from 'react';

export interface DashboardProps {
  title: string;
  dashBoardStyles?: React.CSSProperties;
  children: React.ReactNode;
}

const DashBoard: React.FC<DashboardProps> = ({ title, dashBoardStyles, children }) => {
  return (
    <Box
      maxHeight={'500px'}
      minHeight={'300px'}
      minWidth={'200px'}
      sx={{
        ...dashBoardStyles,
        overflow: 'auto',
      }}
      padding={'0.5rem'}
    >
      <Grid2 container spacing={2} paddingTop={'0.5rem'}>
        <Grid2 size={12}>
          <Box
            height={'auto'}
            paddingX={'0.5rem'}
            paddingBottom={'0.75rem'}
            display={'flex'}
            justifyContent={'space-between'}
          >
            <Typography variant='h5'>{title}</Typography>
          </Box>
          <Divider />
        </Grid2>
        <Grid2 container width={'100%'} display={'flex'} justifyContent={'space-between'}>
          {children}
        </Grid2>
      </Grid2>
    </Box>
  );
};

export default DashBoard;
