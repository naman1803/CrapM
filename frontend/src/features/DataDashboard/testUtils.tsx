jest.mock('recharts', () => ({
  __esModule: true,
  ...jest.requireActual('recharts'),
  ResponsiveContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>ResponsiveContainer</p>
      {children}
    </div>
  )),
}));
