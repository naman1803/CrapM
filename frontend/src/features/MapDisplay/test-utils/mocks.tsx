import {
  LakeGeoPoint,
  MapElement,
  MapElementType,
  SewageGeoPoint,
} from '../types/MapElements';

export const mockLakeElements: MapElement[] = [
  {
    id: 1,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      name: 'Grrrr Lake',
      additional_notes: 'This is a lake',
    } as LakeGeoPoint,
  },
  {
    id: 2,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      name: 'Obama Lake',
    } as LakeGeoPoint,
  },
];

export const mockSewageElements: MapElement[] = [
  {
    id: 1,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      facility_owner: 'Eric',
      radius: 100,
      colour: 'Blue',
    } as SewageGeoPoint,
  },
  {
    id: 2,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      facility_owner: 'Myron',
      radius: 100,
      colour: 'Red',
    } as SewageGeoPoint,
  },
];
