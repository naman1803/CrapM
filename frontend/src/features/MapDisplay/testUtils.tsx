import { LeafletEventHandlerFnMap, LatLngExpression } from 'leaflet';

jest.mock('react-leaflet', () => ({
  __esModule: true,
  ...jest.requireActual('leaflet'),
  useMap: () => ({
    flyTo: jest.fn(),
    getZoom: jest.fn().mockReturnValue(13), // Add getZoom method returning default zoom level
    on: jest.fn(), // Mock on method for event handling
    off: jest.fn(), // Mock off method for removing event listeners
  }),
  MapContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Map Container</p>
      {children}
    </div>
  )),
  Marker: jest
    .fn()
    .mockImplementation(
      ({
        position,
        children,
        key,
        eventHandlers,
      }: {
        eventHandlers: LeafletEventHandlerFnMap;
        position: LatLngExpression;
        key: any;
        children: React.ReactNode;
      }) => (
        <div key={key}>
          <p>{`Marker-${position.toString()}`}</p>
          {Object.entries(eventHandlers).map(([keyE, value]) => (
            <button key={keyE} onClick={value}>
              {`button-${keyE}-${position.toString()}`}
            </button>
          ))}
          {children}
        </div>
      ),
    ),
  Popup: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Popup</p>
      {children}
    </div>
  )),
  TileLayer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>TileLayer</p>
      {children}
    </div>
  )),
}));
