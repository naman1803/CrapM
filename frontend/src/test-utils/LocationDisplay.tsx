import { useLocation } from 'react-router-dom';
/**test component which displays the current url location.
 * use testid = location-display to get value of current location
 */
const LocationDisplay: React.FC = () => {
  const location = useLocation();

  return <div data-testid='location-display'>{location.pathname}</div>;
};
export default LocationDisplay;
