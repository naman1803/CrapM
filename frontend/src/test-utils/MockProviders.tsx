import { ThemeProvider } from '@mui/material';
import { MemoryRouter } from 'react-router-dom';
import { themeOptions } from '@app/theme';
import React from 'react';

// Wraps all providers for the app for testing
export const TestProviders = ({
  children,
  initialRoute = '/',
}: {
  children?: React.ReactNode;
  initialRoute?: string;
}) => {
  return (
    <ThemeProvider theme={themeOptions}>
      <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
    </ThemeProvider>
  );
};
