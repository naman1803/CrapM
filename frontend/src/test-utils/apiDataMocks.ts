import { LakeData } from '@/types/LakeData';
import {
  SewageCoordData,
  SewageData,
  SewageSeverityData,
} from '@/types/SewageData';

export const mockSewageSeverityElement: SewageSeverityData = {
  id: 1,
  facilityOwner: 'Eric',
  contaminants: {
    brainrot: {
      averageSeverityScore: 1.0,
      overallSeverityLevel: 'Moderate',
      averagePercentExceedance: 100.0,
      records: [
        {
          severityScore: 1.0,
          severityLevel: 'Moderate',
          date: '2021-01-01',
        },
        {
          severityScore: 2.0,
          severityLevel: 'High',
          date: '2021-01-02',
        },
      ],
    },
    takingDumb: {
      averageSeverityScore: 2.0,
      overallSeverityLevel: 'High',
      averagePercentExceedance: 200.0,
      records: [
        {
          severityScore: 0.5,
          severityLevel: 'High',
          date: '2021-01-01',
        },
        {
          severityScore: 2.0,
          severityLevel: 'High',
          date: '2021-01-02',
        },
      ],
    },
  },
};

export const mockLakeApiData: LakeData[] = [
  {
    id: 1,
    lakeName: 'Lake Ontario',
    latitude: 45.2,
    longitude: -78.94,
    additionalNotes: null,
  },
  {
    id: 2,
    lakeName: 'Lake Erie',
    latitude: 46.47,
    longitude: -82.78,
    additionalNotes: null,
  },
];

export const mockSewageApiDataSent = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    latitude: 44.49,
    longitude: -77.69,
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    radius: 22.539892,
    colour: "Green"
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    latitude: 44.68,
    longitude: -79.23,
    contaminant: 'PH - HIGH',
    radius: 22.524761,
    colour: "Green"
  },
];
export const mockSewageApiData: SewageCoordData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    latitude: 44.49,
    longitude: -77.69,
    radius: 22.539892,
    colour: "Green"
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    latitude: 44.68,
    longitude: -79.23,
    radius: 22.524761,
    colour: "Green"
  },
];

export const mockSewageSeverityApiData: SewageSeverityData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    contaminants: {
      'BIOCHEMICAL OXYGEN DEMAND': {
        averageSeverityScore: 0.77,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 177.0,
        records: [
          {
            severityScore: 0.77,
            severityLevel: 'Low',
            date: '2016-08-18',
          },
        ],
      },
    },
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    contaminants: {
      'PH - HIGH': {
        averageSeverityScore: 0.95,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 195.4,
        records: [
          {
            severityScore: 0.95,
            severityLevel: 'Low',
            date: '2016-08-30',
          },
        ],
      },
      'BIOCHEMICAL OXYGEN DEMAND': {
        averageSeverityScore: 0.58,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 157.8,
        records: [
          {
            severityScore: 0.58,
            severityLevel: 'Low',
            date: '2016-08-23',
          },
        ],
      },
    },
  },
];

export const mockSewageAllApiData: SewageData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    siteAddress: '123 Mock St',
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-18T00:00:00',
    exceedanceEndDate: '2016-08-18T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 88.5,
    quantityMinimum: null,
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    siteAddress: '1234 Mock St',
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-23T00:00:00',
    exceedanceEndDate: '2016-08-23T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 78.9,
    quantityMinimum: null,
  },
  {
    id: 3,
    facilityOwner: 'Mock Facility 2',
    siteAddress: '1234 Mock St',
    contaminant: 'PH - HIGH',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-30T00:00:00',
    exceedanceEndDate: '2016-08-30T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 97.7,
    quantityMinimum: null,
  },
];
