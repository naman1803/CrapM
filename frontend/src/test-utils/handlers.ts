import { http, HttpResponse } from 'msw';
import {
  mockLakeApiData,
  mockSewageAllApiData,
  mockSewageApiDataSent,
  mockSewageSeverityApiData,
} from './apiDataMocks';

const handlers = [
  http.get('/api/maps/lakes/all', () => {
    return HttpResponse.json(mockLakeApiData);
  }),
  http.get('/api/maps/sewage/all', () => {
    return HttpResponse.json(mockSewageApiDataSent);
  }),
  http.get('/api/data/sewage/all', () => {
    return HttpResponse.json(mockSewageAllApiData);
  }),
  ...mockSewageSeverityApiData.map((severityData) =>
    http.get(`/api/data/sewage/severity/${severityData.id}`, () => {
      return HttpResponse.json(severityData);
    }),
  ),
];
export default handlers;
