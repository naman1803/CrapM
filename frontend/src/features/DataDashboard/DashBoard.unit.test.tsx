import './testUtils';

import { TestProviders } from '@/test-utils/MockProviders';
import { render } from '@testing-library/react';
import DashBoard, { DashboardProps } from './DashBoard';
import { Typography } from '@mui/material';
import ResizeObserver from 'resize-observer-polyfill';

// resizeobserver is not defined
beforeAll(() => {
    globalThis.ResizeObserver = ResizeObserver;
});

const mockDashBoardData: DashboardProps = {
    title: 'Analytics',
    dashBoardStyles: undefined,
    children: <Typography>Hello World</Typography>
}

// resizeobserver is not defined
beforeAll(() => {
    globalThis.ResizeObserver = class {
        observe() { }
        unobserve() { }
        disconnect() { }
    };
});

describe('Dashboard', () => {
    it('should render the Dashboard', () => {
        const { asFragment } = renderDashboard(mockDashBoardData);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should display the Dashboard title', () => {
        const { getByText } = renderDashboard(mockDashBoardData);
        expect(getByText('Analytics')).toBeInTheDocument();
    });
    it('should display its child components', () => {
        const { getByText } = renderDashboard(mockDashBoardData);
        expect(getByText('Hello World')).toBeInTheDocument();
    });
});

const renderDashboard = (props: DashboardProps) => render(
    <TestProviders>
        <DashBoard {...props} />
    </TestProviders>
);
