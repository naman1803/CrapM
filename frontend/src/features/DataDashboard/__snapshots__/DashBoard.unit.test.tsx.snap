// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render the Dashboard 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-fe63hs"
  >
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row MuiGrid2-spacing-xs-2 css-1fph5u6-MuiGrid2-root"
    >
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
      >
        <div
          class="MuiBox-root css-1vin698"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1niuvds-MuiTypography-root"
          >
            Analytics
          </h5>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-x148m1-MuiDivider-root"
        />
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-1x0rc6b-MuiGrid2-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-tckkrt-MuiTypography-root"
        >
          Hello World
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
